### DISCUSSION TITLE
Idiomatic conversion from sync query to async select

### DISCUSSION THREAD

**Discussion by jgould22** [sqlalchemy/sqlalchemy#7406] [Usage Questions] [answered]
I am converting some synchronous code and want to make sure the queries I am converting are idiomatic

I have an example here that I want to change to async

```python
first_result = (
    session.query(TableA)
    .order_by(TableA.id.desc())
    .first()
)
return first_result.some_column_value
```
What is the most idiomatic way to convert this?

I have made some attempts below but they all seem a lot more clumsy than the initial code (perhaps that is just the cost of doing business with the async api?)

Attempt A
```python
statement = select(
    TableA.id,
).order_by(TableA.id.desc())

request = await session.execute(statement)

first_result =request.first()

return first_result.some_column_value
```

Attempt B
```python
result = (
    await session.execute(
        select(
            TableA.id,
        ).order_by(TableA.id.desc())
    )
).first()

return result.some_column_value
```

Additional question to do with the above, calling `first()` on the query in the sync version would set the limit to 1 and return only one result, right?. 

However with the async code I am spawning the co-routine and then calling first on the AsycResult.

Is it getting all the results and just grabbing the first when I call `first()` or is it setting the limit like the query object and only getting a single value ?

**Comment by CaselIT** [comment, marked as answer]
Hi,

You can probably look at https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-orm-usage for examples of common patterns.

What you would probably do in this case is
```py
stmt = select(TableA).order_by(TableA.id.desc()).limit(1)
first_result = (await session.scalars(stmt)).first()
return first_result.some_column_value
```

Selecting `select(TableA.id)` will return only the `id` column, not instances of the model `TableA`

> Is it getting all the results and just grabbing the first when I call first() or is it setting the limit like the query object and only getting a single value ?

the fist one, you should manually add `limit(1)`. query did that for you mostly, but it would break in some edge cases and was in general not very reliable with complex queries, so v2 opted for explicit is better

**Reply by jgould22** [reply]
Thank you for the quick reply!

`Table.id` is a typo from trimming down the actual query which has a large number of specific columns but not all of them but I understand what you are saying.

I had not seen that chart but it is very helpful thank you.

Is the example missing the `= await` on the second line?

**Reply by CaselIT** [reply]
> Is the example missing the = await on the second line?

it did. I've updated it

**Reply by jgould22** [reply]
Ah I cant seem to use 
``` python
first_result = await session.scalars(stmt).first()
```
since it causes an `E   AttributeError: 'coroutine' object has no attribute 'first'`

Seems I have to await the coroutine and the call first on the result so the brackets or second line are required

**Reply by CaselIT** [reply]
yes, brackets were missing. updated
